### DISCUSSION TITLE
Foreground

### DISCUSSION THREAD

**Discussion by hxhgxy** [MaterialDesignInXAML/MaterialDesignInXamlToolkit#3383] [Q&A]
Can I ask what is the difference between two Foreground for text below?
```
Property="Foreground"
			Value="{Binding RelativeSource={RelativeSource AncestorType={x:Type FrameworkElement}}, Path=(TextElement.Foreground)}"
```

```
Property="Foreground"
			Value="{DynamicResource MaterialDesign.Brush.Foreground}"
```

Would they have different behavior in Theme changing? Light <---> Dark

**Comment by nicolaihenriksen** [comment]
@hxhgxy I'll try to provide an answer:

```xaml
Property="Foreground" Value="{DynamicResource MaterialDesign.Brush.Foreground}"
```
This will explicitly set the foreground to brush defined by the `MaterialDesign.Brush.Foreground`, and since it is using `DynamicResource` it will update if the theme is changed in runtime.

```xaml
Property="Foreground" Value="{Binding RelativeSource={RelativeSource AncestorType={x:Type FrameworkElement}}, Path=TextElement.Foreground)}"
```
This will bind the foreground to the brush which is set on the nearest parent element of type `FrameworkElement`. You may or may not get the same behavior as above, depending on how the brush is set on the parent element. If a `DynamicResource` like above is used on the parent element, the result will be the same. However, if for some reason, the foreground brush on the parent element was explicitly set to something else (e.g. `Foreground=Fuchsia`) then that is the color you will bind to (and in this particular "hardcoded" case it will not update if the theme is changed in runtime).

The latter approach will typically be used in places where the control/element you're currently setting properties for does not really know (or care) about the context it resides in, but it just needs to "follow along".

And it may be worth pointing out that `TextElement.Foreground` is an inheritable dependency property. So unless it is explicitly set using `TextElement.Foreground=[something]` on your element, it will effectively inherit the value of the nearest parent that has it explicitly set.

**Comment by hxhgxy** [comment]
Thank you! I try to learn and understand your codes on theme management. It is still too difficult to me. 
If I use DynamicResource MD.Brush.Foreground, the foreground should change immediately once I switch the basetheme, right?

**Reply by nicolaihenriksen** [reply]
@hxhgxy Yes, the `DynamicResource` is designed for this particular scenario where the resource for a given key is replaced during runtime; which is exactly what happens when changing theme.
